design/one-pager-k8s-native-providers.md

```diff
@@ -0,0 +1,217 @@
+
+```yaml
+apiVersion: database.crossplane.io/v1alpha1
+kind: PostgreSQLInstance
+metadata:
+  name: app-postgresql
+  namespace: app-project1-dev
+spec:
+  classRef:
+    name: postgresql-standard
+  writeConnectionSecretToRef:
+    name: postgresqlconn
+  engineVersion: "9.6"
+```
+
+On creation of the `PostgreSQLInstance` claim, the claim controller would
+configure and create a `Cluster` (`cockroachdb.rook.crossplane.io/v1alpha1`)
+instance. Then, the managed controller would use the Crossplane `Cluster`
+instance to create a `Cluster` (`cockroachdb.rook.io/v1alpha1`) in the target
+Kubernetes cluster where Rook and the CockroachDB operator are installed. This
+workflow is similar to dynamic provisioning in a typical infrastructure
+provider:
+
+![K8s Native Providers](./images/k8s-native-providers.png)
+
+## Technical Design
+
+### Claim Reconciler
+
+The claim controllers could use the same shared [claim reconciler] that other
+existing infrastructure stacks use currently, providing their own
+`ManagedConfigurators` to configure the managed resource instance using the
+referenced class.
+
+### Scheduler Reconciler
+
+Because the scheduling behavior for most Kubernetes-native infrastructure
+components will look very similar, a shared scheduler reconciler should be added
+to [crossplane-runtime]. It will look almost identical to the reconciler used
+for `KubernetesApplication` [scheduler controller]. In fact, the
+`KubernetesApplication` scheduler controller should likely make use of the
+shared `crossplane-runtime` scheduler reconciler after it is implemented.
+
+### Managed Reconciler
+
+Though this design is motivated by the desire to support Rook, the pattern can
+be generalized to be applicable to other Kubernetes-native infrastructure
+providers. While most existing infrastructure provider stacks must create a
+client wrapper library to use the CRUD functions of the provider API, every
+Kubernetes-native infrastructure stack will use the [client-go] SDK for CRUD
+operations. Therefore, the pluggable `Observe`, `Create`, `Update`, `Delete`
+functions in the [crossplane-runtime] can likely have a common implementation
+for Kubernetes-native infrastructure stacks.
+
+This can be done via generic `NativeConnecter` and `nativeExternal` types in
+`crossplane-runtime`. Similar to the `clusterConnecter` in the
+`KubernetesApplicationResource` controller, it would get the cluster connection
+information of the Kubernetes cluster that the managed resource was scheduled to
+by the scheduler controller, then return a `nativeExternal` type that implements
+the `ExternalClient` interface. Because external client functions will
+essentially be making sure the "external resource" in the target cluster matches
+the "managed resource" (i.e. they should be nearly identical Kubernetes
+resources), it should be possible to have generic implementations of the
+`ExternalClient` functions. However, if it proves difficult to do so, or it just
```

Comment:
All current implementations of `resource.ExternalClient` type assert the `resource.Managed` they're supplied with to a concrete type. This is primarily because each type we deal with has different `Spec` and `Status` structs that we must translate to cloud provider API types.

Do you imagine that (for example) `cockroachdb.rook.crossplane.io/v1alpha1` would be a strict superset of `cockroachdb.rook.io/v1alpha1` to the extent we could have the `nativeExternal` `resource.ExternalClient` treat the "external resources" as `*unstructured.Unstructured` or similar? If so, do you have an idea of how we'd assess the readiness of such external resources so that we could update our `BindingPhase` and `Ready` condition? `KubernetesApplicationResource` basically punts on this by calling the opaque external resource `Submitted` rather than `Ready`.

Reply:
In the general case, we could just check that the resource exists in the remote cluster. However, if the external client was manually implemented then we could do more intelligent checking of any `Status` fields that the "provider" surfaces on their resources.